File: 2024/day-18/main.lua
local function parse_input(file_name)
    local points = {}
    for line in io.lines(file_name) do
        local x, y = line:match("(%d+),(%d+)")
        x, y = tonumber(x), tonumber(y)
        table.insert(points, {x = x, y = y})
    end
    return points
end

local function build_grid(size)
    local grid = {}
    for i = 1, size do
        grid[i] = {}
        for j = 1, size do
            grid[i][j] = false
        end
    end
    return grid
end

local function bfs(grid, start, goal)
    local size = #grid
    local directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}}
    local queue = {{x = start.x, y = start.y}}
    local visited = {}
    visited[start.y * size + start.x] = true

    while #queue > 0 do
        local current = table.remove(queue, 1)
        if current.x == goal.x and current.y == goal.y then
            return true
        end

        for _, dir in ipairs(directions) do
            local nx, ny = current.x + dir[1], current.y + dir[2]
            if nx >= 1 and nx <= size and ny >= 1 and ny <= size and
               not grid[ny][nx] and not visited[ny * size + nx] then
                table.insert(queue, {x = nx, y = ny})
                visited[ny * size + nx] = true
            end
        end
    end
    return false
end

local function find_shortest_path(grid, start, goal)
    local size = #grid
    local directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}}
    local queue = {{x = start.x, y = start.y, steps = 0}}
    local visited = {}
    visited[start.y * size + start.x] = true

    while #queue > 0 do
        local current = table.remove(queue, 1)
        if current.x == goal.x and current.y == goal.y then
            return current.steps
        end

        for _, dir in ipairs(directions) do
            local nx, ny = current.x + dir[1], current.y + dir[2]
            if nx >= 1 and nx <= size and ny >= 1 and ny <= size and
               not grid[ny][nx] and not visited[ny * size + nx] then
                table.insert(queue, {x = nx, y = ny, steps = current.steps + 1})
                visited[ny * size + nx] = true
            end
        end
    end
    return -1
end

local function main()
    local file_name = "input.txt"
    local points = parse_input(file_name)
    local size = 71
    local grid = build_grid(size)

    local start = {x = 1, y = 1}
    local goal = {x = size, y = size}

    -- Part 1: Shortest path after 1024 bytes
    for i = 1, math.min(1024, #points) do
        local point = points[i]
        grid[point.y + 1][point.x + 1] = true
    end
    local shortest_path = find_shortest_path(grid, start, goal)
    if shortest_path == -1 then
        print("No valid path to the exit after 1024 bytes.")
    else
        print("Shortest Path after 1024 bytes: " .. shortest_path)
    end

    -- Reset grid for Part 2
    grid = build_grid(size)

    -- Part 2: First byte that blocks the path
    for i, point in ipairs(points) do
        grid[point.y + 1][point.x + 1] = true
        if not bfs(grid, start, goal) then
            print(string.format("First blocking byte: %d,%d", point.x, point.y))
            break
        end
    end
end

main()

File: 2024/day-18/main.rs
use std::collections::{HashSet, VecDeque};
use std::fs;

type Point = (usize, usize);

fn parse_input(file_name: &str) -> Vec<Point> {
    let input = fs::read_to_string(file_name).expect("Failed to read input file");
    input
        .lines()
        .map(|line| {
            let coords: Vec<usize> = line.split(',').map(|x| x.parse().unwrap()).collect();
            (coords[0], coords[1])
        })
        .collect()
}

fn is_within_bounds(x: isize, y: isize, size: usize) -> bool {
    x >= 0 && x < size as isize && y >= 0 && y < size as isize
}

fn bfs(grid: &HashSet<Point>, start: Point, goal: Point, size: usize) -> bool {
    let directions = [(0, 1), (1, 0), (0, -1), (-1, 0)];
    let mut queue = VecDeque::new();
    let mut visited = HashSet::new();

    queue.push_back(start);
    visited.insert(start);

    while let Some((x, y)) = queue.pop_front() {
        if (x, y) == goal {
            return true;
        }

        for &(dx, dy) in &directions {
            let nx = x as isize + dx;
            let ny = y as isize + dy;
            if is_within_bounds(nx, ny, size) {
                let neighbor = (nx as usize, ny as usize);
                if !grid.contains(&neighbor) && visited.insert(neighbor) {
                    queue.push_back(neighbor);
                }
            }
        }
    }
    false
}

fn find_shortest_path(grid: &HashSet<Point>, start: Point, goal: Point, size: usize) -> Option<usize> {
    let directions = [(0, 1), (1, 0), (0, -1), (-1, 0)];
    let mut queue = VecDeque::new();
    let mut visited = HashSet::new();

    queue.push_back((start, 0));
    visited.insert(start);

    while let Some(((x, y), steps)) = queue.pop_front() {
        if (x, y) == goal {
            return Some(steps);
        }

        for &(dx, dy) in &directions {
            let nx = x as isize + dx;
            let ny = y as isize + dy;
            if is_within_bounds(nx, ny, size) {
                let neighbor = (nx as usize, ny as usize);
                if !grid.contains(&neighbor) && visited.insert(neighbor) {
                    queue.push_back((neighbor, steps + 1));
                }
            }
        }
    }
    None
}

fn main() {
    let file_name = "input.txt";
    let points = parse_input(file_name);
    let size = 71;
    let start = (0, 0);
    let goal = (70, 70);

    // Part 1: Shortest path after 1024 bytes
    let mut grid = HashSet::new();
    for i in 0..1024.min(points.len()) {
        grid.insert(points[i]);
    }
    if let Some(shortest_path) = find_shortest_path(&grid, start, goal, size) {
        println!("Shortest Path after 1024 bytes: {}", shortest_path);
    } else {
        println!("No valid path to the exit after 1024 bytes.");
    }

    // Part 2: First byte that blocks the path
    grid.clear();
    for (i, &point) in points.iter().enumerate() {
        grid.insert(point);
        if !bfs(&grid, start, goal, size) {
            println!("First blocking byte: {},{}", point.0, point.1);
            break;
        }
    }
}
